File: ios/FLEXR/Core/Locale/AppLanguage.swift
import Foundation

/// Die beiden Sprachen der App.
///
/// Deutsch ist die Ausgangssprache: [FlexrStrings] trägt den deutschen Text als
/// Original, Englisch als Übersetzung. Fehlt ein englischer Eintrag, fällt die
/// Suche auf den deutschen zurück statt auf den nackten Schlüssel — eine
/// vergessene Übersetzung sieht dann nach deutschem Text aus und nicht nach
/// einem Fehler.
enum AppLanguage: String, CaseIterable, Sendable {
    case german = "de"
    case english = "en"

    /// Kurzform für den Sprachregler.
    var badge: String { rawValue.uppercased() }

    /// Zeitzonen des deutschsprachigen Raums. Büsingen (deutsche Exklave in der
    /// Schweiz) und Vaduz (Liechtenstein) sind eigene IANA-Zonen und fielen
    /// sonst als „nicht DACH" durch.
    private static let dachZones: Set<String> = [
        "Europe/Vienna", "Europe/Berlin", "Europe/Zurich",
        "Europe/Busingen", "Europe/Vaduz",
    ]

    private static let dachRegions: Set<String> = ["AT", "DE", "CH", "LI"]

    /// Sprache beim ersten Start, solange der Nutzer nichts gewählt hat.
    ///
    /// Vorgabe ist der Standort („DACH-Raum → Deutsch"), nicht die
    /// Systemsprache: ein in Wien gekauftes Gerät mit englischer Systemsprache
    /// steht trotzdem auf Europe/Vienna. Erst wenn der Ort nichts hergibt,
    /// entscheidet die Systemsprache — und wer sein Gerät auf Deutsch gestellt
    /// hat, bekommt Deutsch auch aus Mailand. Englisch ist der Rückfall für
    /// alles Übrige.
    ///
    /// Dieselbe Reihenfolge wie in der Web-App (`frontend/i18n.js`) und in der
    /// Android-Fassung (`AppLanguage.detect`).
    static func detect(
        timeZoneIdentifier: String = TimeZone.current.identifier,
        locale: Locale = .current
    ) -> AppLanguage {
        if dachZones.contains(timeZoneIdentifier) { return .german }
        if let region = locale.region?.identifier.uppercased(), dachRegions.contains(region) {
            return .german
        }
        if locale.language.languageCode?.identifier.lowercased() == "de" { return .german }
        return .english
    }
}

File: android-native/app/src/main/java/flexr/social/app/core/locale/AppLanguage.kt
package flexr.social.app.core.locale

import java.util.Locale
import java.util.TimeZone

/**
 * Die beiden Sprachen der App.
 *
 * Deutsch ist die Ausgangs- und Standardsprache: `res/values/strings.xml`
 * traegt die deutschen Texte, `res/values-en/strings.xml` die englischen.
 * Fehlt eine englische Uebersetzung, faellt Android von selbst auf die
 * deutsche Fassung zurueck.
 */
enum class AppLanguage(val code: String, val locale: Locale) {
    GERMAN("de", Locale.forLanguageTag("de-AT")),
    ENGLISH("en", Locale.ENGLISH),
    ;

    companion object {

        val DEFAULT = GERMAN

        fun fromCode(code: String?): AppLanguage? =
            entries.firstOrNull { it.code == code }

        /**
         * Zeitzonen des deutschsprachigen Raums. Buesingen (deutsche Exklave in
         * der Schweiz) und Vaduz (Liechtenstein) sind eigene IANA-Zonen und
         * wuerden sonst als "nicht DACH" durchfallen.
         */
        private val DACH_ZONES = setOf(
            "Europe/Vienna", "Europe/Berlin", "Europe/Zurich",
            "Europe/Busingen", "Europe/Vaduz",
        )

        private val DACH_COUNTRIES = setOf("AT", "DE", "CH", "LI")

        /**
         * Sprache beim ersten Start, wenn der Nutzer noch nichts gewaehlt hat.
         *
         * Vorgabe ist der Standort ("DACH-Raum -> Deutsch"), nicht die
         * Systemsprache: ein in Wien gekauftes Geraet mit englischer
         * Systemsprache steht trotzdem auf Europe/Vienna. Erst wenn der Ort
         * nichts hergibt, entscheidet die Systemsprache — und wer sein Geraet
         * auf Deutsch gestellt hat, bekommt Deutsch auch aus Mailand.
         * Englisch ist der Rueckfall fuer alles Uebrige.
         *
         * Dieselbe Reihenfolge wie in der Web-App (frontend/i18n.js).
         */
        fun detect(
            timeZoneId: String = TimeZone.getDefault().id,
            deviceLocale: Locale = Locale.getDefault(),
        ): AppLanguage {
            if (timeZoneId in DACH_ZONES) return GERMAN
            if (deviceLocale.country.uppercase() in DACH_COUNTRIES) return GERMAN
            if (deviceLocale.language.lowercase() == "de") return GERMAN
            return if (deviceLocale.language.isBlank()) DEFAULT else ENGLISH
        }
    }
}
